src/network/networkd-address.c

```diff
@@ -20,6 +21,85 @@
 #define ADDRESSES_PER_LINK_MAX 2048U
 #define STATIC_ADDRESSES_PER_NETWORK_MAX 1024U
 
+/* Unique Local IPv6 Unicast Addresses https://tools.ietf.org/rfc/rfc4193.txt */
+int generate_ipv6_ula_address(Link *link, uint16_t subnet_id, Address **ret) {
+        _cleanup_(address_freep) Address *address = NULL;
+        _cleanup_(khash_unrefp) khash *h = NULL;
+        uint8_t key[16], eui64[8], global_id[5];
+        const void *p;
+        uint64_t ntp;
+        int r;
+
+        r = address_new(&address);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not allocate address: %m");
+
+        address->family = AF_INET6;
+
+        /* 1. Obtain the current time of day in 64-bit NTP format [NTP]. */
+        ntp = (int64_t) now(CLOCK_MONOTONIC);
+        memcpy((uint8_t *) &key, &ntp, sizeof(uint64_t));
+
+        /* 2) Get EUI-64 from the interface specified */
+        eui64[0]  = link->mac.ether_addr_octet[0];
+        eui64[0] ^= 1 << 1;
+        eui64[1]  = link->mac.ether_addr_octet[1];
+        eui64[2]  = link->mac.ether_addr_octet[2];
+        eui64[3]  = 0xff;
+        eui64[4]  = 0xfe;
+        eui64[5]  = link->mac.ether_addr_octet[3];
+        eui64[6]  = link->mac.ether_addr_octet[4];
+        eui64[7]  = link->mac.ether_addr_octet[5];
+
+        /*  3) Concatenate the time of day with the system-specific identifier
+         *   in order to create a key.
+         */
+        memcpy((uint8_t *) &key + sizeof(uint64_t), (void *) &eui64, sizeof(eui64));
```

Comment:
Seems odd to first copy the data into the eui64 array, just to copy it into the key array, and then never use it again.

Why not write the data directly into the key array?